File: backend/alere/views/quotes.py
from .json import nan_enc, JSONView
from typing import List
import alere.models
import datetime
import math


class Position:
    def __init__(self, r: alere.models.RoI = None):
        if r is not None:
            self.avg_cost = r.average_cost
            self.equity = r.balance or 0
            self.gains = r.realized_gain or 0
            self.invested = r.invested
            self.pl = r.pl
            self.roi = r.roi
            self.shares = r.shares
            self.weighted_avg = r.weighted_average
        else:
            self.avg_cost = math.nan
            self.equity = 0
            self.gains = 0
            self.invested = 0
            self.pl = 0
            self.roi = math.nan
            self.shares = 0
            self.weighted_avg = math.nan

    def to_json(self):
        return {
            "avg_cost": nan_enc(self.avg_cost),
            "equity": nan_enc(self.equity),
            "gains": nan_enc(self.gains),
            "invested": nan_enc(self.invested),
            "pl": nan_enc(self.pl),
            "roi": nan_enc(self.roi),
            "shares": nan_enc(self.shares),
            "weighted_avg": nan_enc(self.weighted_avg),
        }

    def __repr__(self):
        return str(self.to_json())


class Account:
    """
    Details on an investment account
    """
    def __init__(self, account_id):
        self.account_id = account_id
        self.start = Position()  # as of mindate
        self.end = Position()    # as of maxdate
        self.oldest = None       # date of oldest transaction (for annualized)
        self.most_recent = None  # most recent transaction
        self.prices = []

    def to_json(self):
        now = datetime.datetime.now().astimezone(datetime.timezone.utc)

        # Annualized total return on investment
        if self.oldest and self.end.roi:
            d = (now - self.oldest).total_seconds() / 86400
            annualized_return = self.end.roi ** (365 / d)
        else:
            annualized_return = math.nan

        # Return over the period [mindate,maxdate]
        d = self.start.equity + self.end.invested - self.start.invested
        period_roi = (
            math.nan if abs(d) < 1e-6
            else (self.end.equity + self.end.gains - self.start.gains) / d
        )

        return {
            "account": self.account_id,
            "start": self.start,
            "end": self.end,
            "oldest": self.oldest.timestamp()
                if self.oldest is not None else 0,
            "most_recent": self.most_recent.timestamp()
                if self.most_recent is not None else 0,
            "now_for_annualized": now.timestamp(),
            "annualized_roi": nan_enc(annualized_return),
            "period_roi": nan_enc(period_roi),
            "prices": self.prices,
        }

    def __repr__(self):
        return str(self.to_json())


class Symbol:
    """
    Details on a traded symbol
    """
    def __init__(
            self,
            commodity_id: int,
            ticker: str,
            source: int,
            is_currency: bool,
            price_scale: int
        ):
        self.id = commodity_id
        self.source = source
        self.ticker = ticker
        self.accounts: List[alere.models.Accounts] = []   # array of Account
        self.oldest = None
        self.is_currency = is_currency
        self.price_scale = price_scale

    def to_json(self):
        return {
            "id": self.id,
            "ticker": self.ticker,
            "source": self.source,
            "accounts": self.accounts,
            "is_currency": self.is_currency,
        }

    def __repr__(self):
        return str(self.to_json())


class QuotesView(JSONView):

    def get_json(self, params):
        currency = self.as_commodity_id(params, 'currency')
        commodities = params.get('commodities', None)
        accounts = params.get('accounts')  # comma-separate list of ids
        maxdate = self.as_time(params, 'maxdate')
        mindate = self.as_time(params, 'mindate')

        if accounts:
            accounts = [int(c) for c in accounts.split(',')]

        #########
        # Find all commodities

        query = alere.models.Commodities.objects.all()
        if commodities:
            query = query.filter(
                id__in=[int(c) for c in commodities.split(',')])

        symbols = {
            c.id: Symbol(
                commodity_id=c.id,
                ticker=c.quote_symbol,
                source=c.quote_source_id or alere.models.PriceSources.USER,
                is_currency=c.kind == alere.models.CommodityKinds.CURRENCY,
                price_scale=c.price_scale,
            )
            for c in query
        }

        ########
        # Find the corresponding accounts

        accs = {}

        query = alere.models.Accounts.objects \
            .filter(kind__is_trading=True) \
            .select_related('commodity')
        if accounts is not None:
            query = query.filter(id__in=accounts)

        for row in query:
            a = accs[row.id] = Account(row.id)
            symbols[row.commodity_id].accounts.append(a)

        # remove all symbols for which we have zero account, to limit the scope
        # of the following query
        symbols = {s: data for s, data in symbols.items() if data.accounts}

        ########
        # Compute metrics

        query2 = alere.models.RoI.objects \
            .filter(account_id__in=accs.keys(),
                    currency_id=currency) \
            .order_by("mindate")

        for r in query2:
            # ??? prices might be given in a currency other than currency
            # They could also each use a different currency

            a = accs[r.account_id]

            if a.oldest is None:
                a.oldest = r.mindate
            a.most_recent = r.mindate

            if r.mindate <= mindate and mindate < r.maxdate:
                a.start = Position(r)

            if r.mindate <= maxdate and maxdate < r.maxdate:
                a.end = Position(r)

            a.prices.append({
                "t": r.mindate.timestamp() * 1e3,
                "price": r.computed_price,
                "roi": (r.roi - 1) * 100 if r.roi else None,
                "shares": r.shares,
            })

        return list(symbols.values())

File: alere_lib/src/quotes.rs
use crate::account_lists::price_sources;
use crate::commodities::Commodity;
use crate::connections::SqliteConnect;
use crate::errors::AlrResult;
use crate::models::{AccountId, CommodityId};
use chrono::{DateTime, NaiveDateTime, TimeZone, Utc};
use diesel::prelude::*;
use diesel::sql_types::{Float, Integer, Nullable, Timestamp};
use log::info;
use serde::Serialize;
use std::collections::HashMap;

#[derive(QueryableByName, Debug)]
struct Roi {
    #[diesel(sql_type = Integer)]
    pub account_id: AccountId,

    #[diesel(sql_type = Integer)]
    #[diesel(column_name = commodity_id)]
    pub _commodity_id: CommodityId,

    #[diesel(sql_type = Timestamp)]
    pub min_ts: NaiveDateTime,

    #[diesel(sql_type = Timestamp)]
    pub max_ts: NaiveDateTime,

    #[diesel(sql_type = Nullable<Float>)]
    pub shares: Option<f32>,

    #[diesel(sql_type = Nullable<Float>)]
    pub realized_gains: Option<f32>,

    #[diesel(sql_type = Nullable<Float>)]
    pub price_paid: Option<f32>,

    #[diesel(sql_type = Nullable<Float>)]
    pub shares_worth: Option<f32>,

    #[diesel(sql_type = Nullable<Float>)]
    pub computed_price: Option<f32>,

    #[diesel(sql_type = Nullable<Float>)]
    pub pl: Option<f32>,

    #[diesel(sql_type = Nullable<Float>)]
    pub roi: Option<f32>,

    #[diesel(sql_type = Nullable<Float>)]
    pub weighted_average: Option<f32>,

    #[diesel(sql_type = Nullable<Float>)]
    pub average_cost: Option<f32>,
}

#[derive(Serialize)]
pub struct Position {
    avg_cost: f32,
    equity: f32,
    gains: f32,
    invested: f32,
    pl: f32,
    roi: f32,
    shares: f32,
    weighted_avg: f32,
}

impl Position {
    fn new(roi: &Roi) -> Self {
        Position {
            avg_cost: roi.average_cost.unwrap_or(f32::NAN),
            equity: roi.shares_worth.unwrap_or(f32::NAN),
            gains: roi.realized_gains.unwrap_or(f32::NAN),
            invested: roi.price_paid.unwrap_or(f32::NAN),
            pl: roi.pl.unwrap_or(f32::NAN),
            roi: roi.roi.unwrap_or(f32::NAN),
            shares: roi.shares.unwrap_or(f32::NAN),
            weighted_avg: roi.weighted_average.unwrap_or(f32::NAN),
        }
    }
}

impl Default for Position {
    fn default() -> Self {
        Position {
            avg_cost: 0.0,
            equity: 0.0,
            gains: 0.0,
            invested: 0.0,
            pl: 0.0,
            roi: 0.0,
            shares: 0.0,
            weighted_avg: 0.0,
        }
    }
}

#[derive(Serialize)]
pub struct Price {
    t: i64, //  DateTime<Utc>,
    price: f32,
    roi: f32,
    shares: f32,
}

/// Details on an investment account

#[derive(Serialize)]
pub struct ForAccount {
    account: AccountId,
    start: Position,                    // as of min_ts
    end: Position,                      // as of max_ts
    oldest: Option<DateTime<Utc>>,      // oldest transaction (for annualized)
    most_recent: Option<DateTime<Utc>>, // most recent transaction
    now_for_annualized: DateTime<Utc>,
    prices: Vec<Price>,
    annualized_roi: f32,
    period_roi: f32,
}

impl ForAccount {
    pub fn new(id: AccountId) -> ForAccount {
        ForAccount {
            account: id,
            start: Default::default(),
            end: Default::default(),
            oldest: None,
            most_recent: None,
            now_for_annualized: Utc::now(),
            prices: vec![],
            annualized_roi: f32::NAN,
            period_roi: f32::NAN,
        }
    }
}

/// Details on a traded symbol

#[derive(Serialize)]
pub struct Symbol {
    //  <'a> {
    id: CommodityId,
    ticker: String,
    source: i32,
    is_currency: bool,
    accounts: Vec<AccountId>,
    price_scale: i32,
}

#[derive(QueryableByName)]
struct AccountIdAndCommodity {
    #[diesel(sql_type = Integer)]
    id: AccountId,

    #[diesel(sql_type = Integer)]
    commodity_id: CommodityId,
}

pub fn quotes(
    mut connection: SqliteConnect,
    min_ts: DateTime<Utc>,
    max_ts: DateTime<Utc>,
    currency: CommodityId,
    commodities: Option<Vec<CommodityId>>,
    accounts: Option<Vec<AccountId>>,
) -> AlrResult<(Vec<Symbol>, HashMap<AccountId, ForAccount>)> {
    info!("quotes {:?} {:?} {}", &min_ts, &max_ts, currency);

    // Find all commodities

    let mut all_commodities: Vec<Commodity> = {
        use crate::schema::alr_commodities::dsl::*;
        alr_commodities.load::<Commodity>(&mut connection.0)?
    };

    if let Some(commodities) = commodities {
        all_commodities.retain(|comm| commodities.contains(&comm.id));
    }
    let mut symbols = HashMap::new();
    all_commodities.iter_mut().for_each(|comm| {
        symbols.insert(
            comm.id,
            Symbol {
                id: comm.id,
                ticker: comm.quote_symbol.as_ref().cloned().unwrap_or_default(),
                source: comm.quote_source_id.unwrap_or(price_sources::USER),
                is_currency: comm.is_currency(),
                price_scale: comm.price_scale,
                accounts: vec![],
            },
        );
    });

    // Find the corresponding accounts

    let filter_account = match accounts {
        Some(accs) => format!(
            "AND a.id IN ({})",
            accs.iter()
                .map(|id| id.to_string())
                .collect::<Vec<_>>()
                .join(",")
        ),
        None => "".to_string(),
    };
    let query = format!(
        "SELECT a.id, a.commodity_id \
        FROM alr_accounts a  \
        JOIN alr_account_kinds k ON (a.kind_id = k.id) \
        WHERE k.is_trading {filter_account}"
    );
    let accounts = connection.exec::<AccountIdAndCommodity>("quotes", &query)?;
    let mut accs = HashMap::new();
    accounts.iter().for_each(|a| {
        let acc = ForAccount::new(a.id);
        accs.insert(a.id, acc);
        symbols
            .get_mut(&a.commodity_id)
            .unwrap()
            .accounts
            .push(a.id);
    });

    // Remove all symbols for which we have zero account, to limit the scope
    // of the following query.

    symbols.retain(|_, symb| !symb.accounts.is_empty());

    // Compute metrics

    let accs_ids = accs
        .iter()
        .map(|(id, _)| id.to_string())
        .collect::<Vec<_>>()
        .join(",");

    // ??? Also no point in getting too many points, so we should ignore if
    //    granularity is too small.  Perhaps should get the list of historical
    //    prices from alr_balances_currency instead.
    let query = format!(
        "SELECT * \
        FROM alr_roi r \
        WHERE r.account_id IN ({accs_ids}) \
        AND r.currency_id = {currency} \
        AND r.min_ts < '{max_ts}'
        AND '{min_ts}' < r.max_ts
        ORDER BY r.min_ts"
    );

    let rois = connection.exec::<Roi>("roi", &query)?;
    rois.iter().for_each(|r| {
        let a = accs.get_mut(&r.account_id).unwrap();
        let mi = Utc.from_utc_datetime(&r.min_ts);
        let ma = Utc.from_utc_datetime(&r.max_ts);

        if a.oldest.is_none() {
            // The oldest transaction is also the one that corresponds
            // to the minimal requested date from the user.
            a.oldest = Some(mi);
        }
        if mi <= min_ts && min_ts < ma {
            a.start = Position::new(r);
        }

        a.most_recent = Some(mi);
        a.end = Position::new(r);

        a.prices.push(Price {
            t: mi.timestamp_millis(),
            price: r.computed_price.unwrap_or(f32::NAN),
            roi: match r.roi {
                Some(val) => (val - 1.0) * 100.0,
                None => f32::NAN,
            },
            shares: r.shares.unwrap_or(f32::NAN),
        });
    });

    for (_, a) in accs.iter_mut() {
        let now = Utc::now();

        // Annualized total return on investment
        if let Some(old) = a.oldest {
            let d = (now - old).num_days() as f32;
            a.annualized_roi = f32::powf(a.end.roi, 365.0 / d);
        }

        // Return over the period [mindata, max_ts]
        let d2 = a.start.equity + a.end.invested - a.start.invested;
        if f32::abs(d2) >= 1E-6 {
            a.period_roi = (a.end.equity + a.end.gains - a.start.gains) / d2;
        }
    }

    Ok((symbols.into_values().collect::<Vec<_>>(), accs))
}
